external-crates/move/crates/move-compiler/src/sui_mode/info.rs

```diff
@@ -0,0 +1,304 @@
+    IsUID,
+    /// holds UID directly as one of the fields
+    Direct { field: Field, ty: Loc },
+    /// holds a type which in turn `Direct`ly or `Indirect`ly holds UID
+    Indirect { field: Field, ty: Loc, uid: Loc },
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum TransferKind {
+    /// The object has store
+    PublicTransfer(Loc),
+    /// transferred within the module to an address vis `sui::transfer::transfer`
+    PrivateTransfer(Loc),
+}
+
+#[derive(Debug, Clone)]
+pub struct SuiInfo {
+    /// All types that contain a UID, directly or indirectly
+    /// This requires a DFS traversal of type declarations
+    pub uid_holders: BTreeMap<(ModuleIdent, DatatypeName), UIDHolder>,
+    /// All types that either have store or are transferred privately
+    pub transferred: BTreeMap<(ModuleIdent, DatatypeName), TransferKind>,
+}
+
+impl SuiInfo {
+    pub fn new(
+        pre_compiled_lib: Option<Arc<FullyCompiledProgram>>,
+        modules: &UniqueMap<ModuleIdent, T::ModuleDefinition>,
+        info: &TypingProgramInfo,
+    ) -> Self {
+        assert!(info.sui_flavor_info.is_none());
+        let uid_holders = all_uid_holders(info);
+        let transferred = all_transferred(pre_compiled_lib, modules, info);
+        Self {
+            uid_holders,
+            transferred,
+        }
+    }
+}
+
+/// DFS traversal to find all UID holders
+fn all_uid_holders(info: &TypingProgramInfo) -> BTreeMap<(ModuleIdent, DatatypeName), UIDHolder> {
+    fn merge_uid_holder(u1: UIDHolder, u2: UIDHolder) -> UIDHolder {
+        match (u1, u2) {
+            (u @ UIDHolder::IsUID, _) | (_, u @ UIDHolder::IsUID) => u,
+            (d @ UIDHolder::Direct { .. }, _) | (_, d @ UIDHolder::Direct { .. }) => d,
+            (u1, _) => u1,
+        }
+    }
+
+    fn merge_uid_holder_opt(
+        u1_opt: Option<UIDHolder>,
+        u2_opt: Option<UIDHolder>,
+    ) -> Option<UIDHolder> {
+        match (u1_opt, u2_opt) {
+            (Some(u1), Some(u2)) => Some(merge_uid_holder(u1, u2)),
+            (o1, o2) => o1.or(o2),
+        }
+    }
+
+    // returns true if the type at the given position is a phantom type
+    fn phantom_positions(
+        info: &TypingProgramInfo,
+        sp!(_, tn_): &N::TypeName,
+    ) -> Vec</* is_phantom */ bool> {
+        match tn_ {
+            N::TypeName_::Multiple(n) => vec![false; *n],
+            N::TypeName_::Builtin(sp!(_, b_)) => b_
+                .tparam_constraints(Loc::invalid())
+                .into_iter()
+                .map(|_| false)
+                .collect(),
+            N::TypeName_::ModuleType(m, n) => {
+                let ty_params = match info.datatype_kind(m, n) {
+                    DatatypeKind::Struct => &info.struct_definition(m, n).type_parameters,
+                    DatatypeKind::Enum => &info.enum_definition(m, n).type_parameters,
+                };
+                ty_params.iter().map(|tp| tp.is_phantom).collect()
+            }
+        }
+    }
+
+    fn visit_ty(
```

Comment:
Should this have a `#[growing_stack]` annotation on it? Probably also a good idea to add it to `visit_decl` and `visit_fiels` as well.